--- ports/Mk/Uses/ssl.mk	2020-01-01 11:42:44.000000000 -0500
+++ ports/Mk/Uses/ssl.mk.new	2021-04-14 12:20:51.937225000 -0400
@@ -75,14 +75,6 @@
 	@${ECHO_CMD} "OpenSSL distribution."
 	@${FALSE}
 .  endif
-.  if exists(${LOCALBASE}/lib/libcrypto.so)
-check-depends::
-	@${ECHO_CMD} "Dependency error: This port wants the OpenSSL library from the FreeBSD"
-	@${ECHO_CMD} "base system. You can't build against it, while a newer"
-	@${ECHO_CMD} "version is installed by a port."
-	@${ECHO_CMD} "Please deinstall the port, remove DEFAULT_VERSIONS=ssl=base or undefine WITH_OPENSSL_BASE."
-	@${FALSE}
-.  endif
 
 .else # ${SSL_DEFAULT} != base
 
