--- test-driver.bak	2021-08-04 11:12:04.399976745 +0200
+++ test-driver	2021-08-04 11:17:17.646343985 +0200
@@ -109,7 +109,10 @@
 # to ameliorate tests themselves also writing to the log file. Our tests
 # don't, but others can (automake bug#35762).
 : >"$log_file"
-"$@" >>"$log_file" 2>&1
+case "$1" in
+  *.sh) sh "$@" >>"$log_file" 2>&1 ;;
+  *)    "$@" >>"$log_file" 2>&1 ;;
+esac
 estatus=$?

 if test $enable_hard_errors = no && test $estatus -eq 99; then
