*** xmlversion.in.h.orig
--- xmlversion.in.h
***************
*** 120,126 ****
   *
   * Whether the serialization/saving support is configured in
   */
! #if 1
  #define LIBXML_OUTPUT_ENABLED
  #endif
  
--- 120,126 ----
   *
   * Whether the serialization/saving support is configured in
   */
! #if 0
  #define LIBXML_OUTPUT_ENABLED
  #endif
  
***************
*** 129,135 ****
   *
   * Whether the push parsing interfaces are configured in
   */
! #if 1
  #define LIBXML_PUSH_ENABLED
  #endif
  
--- 129,135 ----
   *
   * Whether the push parsing interfaces are configured in
   */
! #if 0
  #define LIBXML_PUSH_ENABLED
  #endif
  
***************
*** 138,144 ****
   *
   * Whether the xmlReader parsing interface is configured in
   */
! #if 1
  #define LIBXML_READER_ENABLED
  #endif
  
--- 138,144 ----
   *
   * Whether the xmlReader parsing interface is configured in
   */
! #if 0
  #define LIBXML_READER_ENABLED
  #endif
  
***************
*** 156,162 ****
   *
   * Whether the xmlWriter saving interface is configured in
   */
! #if 1
  #define LIBXML_WRITER_ENABLED
  #endif
  
--- 156,162 ----
   *
   * Whether the xmlWriter saving interface is configured in
   */
! #if 0
  #define LIBXML_WRITER_ENABLED
  #endif
  
***************
*** 246,252 ****
   *
   * Whether XPath is configured in
   */
! #if 1
  #define LIBXML_XPATH_ENABLED
  #endif
  
--- 246,252 ----
   *
   * Whether XPath is configured in
   */
! #if 0
  #define LIBXML_XPATH_ENABLED
  #endif
  
***************
*** 255,261 ****
   *
   * Whether XPointer is configured in
   */
! #if 1
  #define LIBXML_XPTR_ENABLED
  #endif
  
--- 255,261 ----
   *
   * Whether XPointer is configured in
   */
! #if 0
  #define LIBXML_XPTR_ENABLED
  #endif
  
***************
*** 273,279 ****
   *
   * Whether iconv support is available
   */
! #if 1
  #define LIBXML_ICONV_ENABLED
  #endif
  
--- 273,279 ----
   *
   * Whether iconv support is available
   */
! #if 0
  #define LIBXML_ICONV_ENABLED
  #endif
  
