*** error.in.h.bak	2023-02-04 18:32:38.885731972 +0100
--- error.in.h	2023-02-04 18:34:04.056640179 +0100
***************
*** 30,35 ****
--- 30,44 ----
  # define _GL_ATTRIBUTE_SPEC_PRINTF_ERROR _GL_ATTRIBUTE_SPEC_PRINTF_SYSTEM
  #endif
  
+ #if GNULIB_REPLACE_ERROR
+ # undef error_print_progname
+ # undef error_message_count
+ # undef error_one_per_line
+ # define error_print_progname rpl_error_print_progname
+ # define error_message_count rpl_error_message_count
+ # define error_one_per_line rpl_error_one_per_line
+ #endif
+ 
  #ifdef __cplusplus
  extern "C" {
  #endif
***************
*** 90,103 ****
  /* If NULL, error will flush stdout, then print on stderr the program
     name, a colon and a space.  Otherwise, error will call this
     function without parameters instead.  */
! extern void (*error_print_progname) (void);
  
  /* This variable is incremented each time 'error' is called.  */
! extern unsigned int error_message_count;
  
  /* Sometimes we want to have at most one error per line.  This
     variable controls whether this mode is selected or not.  */
! extern int error_one_per_line;
  
  #ifdef __cplusplus
  }
--- 99,112 ----
  /* If NULL, error will flush stdout, then print on stderr the program
     name, a colon and a space.  Otherwise, error will call this
     function without parameters instead.  */
! extern DLL_VARIABLE void (*error_print_progname) (void);
  
  /* This variable is incremented each time 'error' is called.  */
! extern DLL_VARIABLE unsigned int error_message_count;
  
  /* Sometimes we want to have at most one error per line.  This
     variable controls whether this mode is selected or not.  */
! extern DLL_VARIABLE int error_one_per_line;
  
  #ifdef __cplusplus
  }
